File: 3312-sorted-gcd-pair-queries/solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
public:
    vector<int> gcdValues(vector<int>& nums, vector<long long>& queries) {
        int max_val = 0;
        for (int x : nums) max_val = max(max_val, x);

        vector<int> freq(max_val + 1, 0);
        for (int x : nums) {
            for (int i = 1; i * i <= x; ++i) {
                if (x % i == 0) {
                    freq[i]++;
                    if (i * i != x) freq[x / i]++;
                }
            }
        }

        vector<long long> count(max_val + 1, 0);
        for (int i = max_val; i >= 1; --i) {
            long long pairs_with_multiple = (1LL * freq[i] * (freq[i] - 1)) / 2;
            for (int j = 2 * i; j <= max_val; j += i) {
                pairs_with_multiple -= count[j];
            }
            count[i] = pairs_with_multiple;
        }

        vector<long long> prefix_sum(max_val + 1, 0);
        for (int i = 1; i <= max_val; ++i) {
            prefix_sum[i] = prefix_sum[i - 1] + count[i];
        }

        vector<int> results;
        for (long long q : queries) {
            auto it = upper_bound(prefix_sum.begin(), prefix_sum.end(), q);
            results.push_back(distance(prefix_sum.begin(), it));
        }
        return results;
    }
};

File: 3312-sorted-gcd-pair-queries/solution.py
from bisect import bisect_right

class Solution:
    def gcdValues(self, nums: list[int], queries: list[int]) -> list[int]:
        max_val = max(nums)
        
        # Count multiples for each potential GCD
        freq = [0] * (max_val + 1)
        for x in nums:
            d = 1
            while d * d <= x:
                if x % d == 0:
                    freq[d] += 1
                    if d * d != x:
                        freq[x // d] += 1
                d += 1
        
        # Calculate exactly how many pairs have gcd == i
        # Using inclusion-exclusion from max_val down to 1
        count = [0] * (max_val + 1)
        for i in range(max_val, 0, -1):
            pairs_with_multiple = (freq[i] * (freq[i] - 1)) // 2
            for j in range(2 * i, max_val + 1, i):
                pairs_with_multiple -= count[j]
            count[i] = pairs_with_multiple
            
        # Build prefix sums to identify rank
        prefix_sum = [0] * (max_val + 1)
        for i in range(1, max_val + 1):
            prefix_sum[i] = prefix_sum[i - 1] + count[i]
            
        # Answer queries using binary search
        results = []
        for q in queries:
            results.append(bisect_right(prefix_sum, q))
            
        return results